### PR TITLE
Percent box

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

File: Applications/Spire/Source/Ui/PercentBox.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+PercentBox::PercentBox(QWidget* parent)
+  : PercentBox(QHash<Qt::KeyboardModifier, Decimal>(
+      {{Qt::NoModifier, 1}, {Qt::AltModifier, 5}, {Qt::ControlModifier, 10},
+      {Qt::ShiftModifier, 20}}), parent) {}
+
+PercentBox::PercentBox(
+    QHash<Qt::KeyboardModifier, Decimal> modifiers, QWidget* parent)
+  : PercentBox(std::make_shared<LocalOptionalDecimalModel>(),
+    std::move(modifiers), parent) {}
```
Comment (kranar) [inline Applications/Spire/Source/Ui/PercentBox.cpp:86]: Indentation.

File: Applications/Spire/Source/Ui/PercentBox.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+PercentBox::PercentBox(std::shared_ptr<OptionalDecimalModel> model,
+    QHash<Qt::KeyboardModifier, Decimal> modifiers, QWidget* parent)
+    : QWidget(parent),
+      m_model(std::move(model)) {
+  m_decimal_box = new DecimalBox(
+    std::make_shared<OptionalPercentModel>(m_model), std::move(modifiers));
+  enclose(*this, *m_decimal_box);
+  setFocusProxy(m_decimal_box);
+  proxy_style(*this, *m_decimal_box);
+  update_style(*m_decimal_box, [] (auto& style) {});
```
Comment (kranar) [inline Applications/Spire/Source/Ui/PercentBox.cpp:97]: I believe line 97 can be removed.

Comment (shaobinspire) [reply Applications/Spire/Source/Ui/PercentBox.cpp:97]: Here line 97 is used to resolve the issue that setting the `TextAlign` of the `PercentBox` to `AlignRight` makes the `DecimalBox` lose the `PaddingRight` style. 
```
      update_style(*PercentBox, [&] (auto& style) {
        style.get(Any()).set(TextAlign(Qt::AlignRight));
      });
```
The update style code above invalidates the style settings in the DecimalBox below:
```
    style.get(+(Any() > is_a<TextBox>()) %
        (is_a<Button>() && matches(Visibility::VISIBLE))).
      set(PaddingRight(scale_width(24)));
```
It is just a workaround. I think there might be something wrong with the styling for the DecimalBox.
Do you know how to style the TextBox in the DecimalBox with the PaddingRight of is 24px when buttons in the DecimalBox are visible? I've spent a long time trying to figure it out, but failed.

Comment (shaobinspire) [reply Applications/Spire/Source/Ui/PercentBox.cpp:97]: I will remove this line and create another task for this issue.

File: Applications/Spire/Source/Ui/PercentBox.cpp [hunk 3]
```diff
@@ -82,8 +82,8 @@ PercentBox::PercentBox(QWidget* parent)
 
 PercentBox::PercentBox(
     QHash<Qt::KeyboardModifier, Decimal> modifiers, QWidget* parent)
-  : PercentBox(std::make_shared<LocalOptionalDecimalModel>(),
-    std::move(modifiers), parent) {}
+    : PercentBox(std::make_shared<LocalOptionalDecimalModel>(),
+      std::move(modifiers), parent) {}
```
Comment (kranar) [inline Applications/Spire/Source/Ui/PercentBox.cpp:86]: Indent line 86 one more level.

**Pull request merged**